### DISCUSSION TITLE
Use of `undefined` instead of null`?

### DISCUSSION THREAD

**Discussion by Codeneos** [mike-lischke/antlr4ng#28] [General]
I noticed that both the original antlr4 js runtime as well as antlt4ng use `null` instead of `undefined` to return nothing everywhere. Working with javascript (and typescript) for sometime it feels somewhat unnatural in the context of these languages.

It also causes small inconveniences when using atnlr4ng, for example when you implement a Visitor you always need to explicitly return null. With undefined you could just not write the `return` statement and it would implicitly return `undefined`. There is more occasions where I find my self writing `XXX ?? undefined` to change null into undefined but it does feel strange that it's required considering the vast majority of TS and JS code prefers undefined over null.

Any change in this would be a breaking change but I was wondering if it is something you would be considering for a major-release?

**Comment by mike-lischke** [comment]
`null` is available in both Java and JS/TS with pretty much the same meaning. It's a natural choice if you don't want to do special checks everywhere a `null` could be used. So, it's easier to use null and with the `!= null`/`== null` non-strict comparison you can easily catch both `null` and `undefined`.

Java doesn't know an `undefined` type and behaves differently in cases where parameters are not specified (which is the only situation which comes close to the use of `undefined` in JS/TS). So for now it's better to stay with `null`, especially as this is a totally valid and useful type in JS/TS.

Personally, I prefer code that is explicit. Makes understanding it easier. The only downside I've seen so far is that you cannot just omit a parameter that cannot be `undefined` and you have to explicitly specify `null`.

I'm not totally against switching to `undefined`, but that would need good justifications and provide some significant benefits for the user.

**Comment by Codeneos** [comment]
First of all, thanks for the extensive reply 👍 

Your right, it is more explicit. But I think that like other java-ports it doesn't use language features such as undefined properly and by default null is preferred when undefined would better match the intent. 

I also don't think it makes sense to just replace all null's with undefined, it would need to be checked on a case by case basis where undefined would better match.

Having said that, everything works perfectly fine with null but it was just something that sprung to mind while using antlr4ng.

But I'm glad your not totally against it, when I get time I'll make some more concrete (and small) proposals and we can discuss if those make sense :)

**Reply by mike-lischke** [reply]
The problem with partial replacement is that you then have to constantly convert between `undefined` and `null`, been there, done that. Which is why I ended up with the current solution, which is sometimes a bit inconvenient, but consistent.
